.hooks/lib/python_checks.sh

```diff
@@ -0,0 +1,199 @@
+#!/usr/bin/env bash
+
+declare -a isort_args=()
+declare -a black_args=()
+: "${fixing:=false}"
+: "${DEFAULT_COVERAGE_FAIL_UNDER:=75}"
+
+setup_python_environment() {
+    local env_label="$1"
+    local project_dir="$2"
+    echo "Setting up ${env_label} environment"
+    uv sync --all-groups --project "$project_dir"
+    # shellcheck disable=SC1091
+    source "$project_dir/.venv/bin/activate"
+    export VIRTUAL_ENV="$project_dir/.venv"
+    export PATH="$VIRTUAL_ENV/bin:$PATH"
+}
+
+load_secondary_scope_paths() {
+    if [ -n "${BLUEOS_SECONDARY_SCOPE:-}" ]; then
+        read -r -a SECONDARY_SCOPE_PATHS <<<"${BLUEOS_SECONDARY_SCOPE}"
+        return
+    fi
+
+    local pyproject="$SECONDARY_PROJECT_DIR/pyproject.toml"
+    if [ ! -f "$pyproject" ]; then
+        SECONDARY_SCOPE_PATHS=()
+        return
+    fi
+
+    local scope_script="$ROOT_DIR/.hooks/lib/get_secondary_scope.py"
+    if [ ! -x "$scope_script" ]; then
+        echo "Missing scope helper: $scope_script"
+        exit 1
+    fi
+
+    mapfile -t SECONDARY_SCOPE_PATHS < <(
+        "$scope_script" --core-dir "$CORE_DIR" --pyproject "$pyproject"
+    )
+}
+
+path_matches_any() {
+    local file="$1"
+    shift || true
+    local scope_path
+    for scope_path in "$@"; do
+        [[ -z "$scope_path" ]] && continue
+        if [[ "$file" == "$scope_path" ]] || [[ "$file" == "$scope_path/"* ]]; then
+            return 0
+        fi
+    done
+    return 1
+}
+
+collect_python_files() {
+    local scope="$1"
+    local -a all_python_files=()
+    mapfile -t all_python_files < <(git -C "$CORE_DIR" ls-files '*.py')
+
+    local -a filtered=()
+    local file
+    for file in "${all_python_files[@]}"; do
+        [[ -z "$file" ]] && continue
+        if path_matches_any "$file" "${SECONDARY_SCOPE_PATHS[@]}"; then
+            if [ "$scope" = "secondary" ]; then
+                filtered+=("$file")
+            fi
+        else
+            if [ "$scope" = "primary" ]; then
+                filtered+=("$file")
+            fi
+        fi
+    done
+
+    printf '%s\n' "${filtered[@]}"
+}
+
+collect_mypy_targets() {
+    local scope="$1"
+    local -a all_projects=()
+    mapfile -t all_projects < <(git -C "$CORE_DIR" ls-files '*/pyproject.toml')
+
+    local -a filtered=()
+    local project
+    for project in "${all_projects[@]}"; do
+        [[ -z "$project" ]] && continue
+        case "$project" in
+            libs/*|services/*)
+                ;;
+            *)
+                continue
+                ;;
+        esac
+
+        if path_matches_any "$project" "${SECONDARY_SCOPE_PATHS[@]}"; then
+            if [ "$scope" = "secondary" ]; then
+                filtered+=("$(dirname "$project")")
+            fi
+        else
+            if [ "$scope" = "primary" ]; then
+                filtered+=("$(dirname "$project")")
+            fi
+        fi
+    done
+
+    printf '%s\n' "${filtered[@]}"
```

Comment:
There is no need to have a filtered field, that makes more complex. If we need a list we can just echo our way around.